paasta_tools/tron_tools.py

```diff
@@ -0,0 +1,379 @@
+import logging
+import os
+from urllib.parse import urljoin
+
+import requests
+import service_configuration_lib
+import yaml
+try:
+    from yaml.cyaml import CSafeDumper as Dumper
+except ImportError:  # pragma: no cover (no libyaml-dev / pypy)
+    Dumper = yaml.SafeDumper
+
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import get_user_agent
+from paasta_tools.utils import InstanceConfig
+from paasta_tools.utils import load_system_paasta_config
+from paasta_tools.utils import load_v2_deployments_json
+
+
+log = logging.getLogger(__name__)
+
+SPACER = '.'
+
+
+class TronNotConfigured(Exception):
+    pass
+
+
+class TronConfig(dict):
+
+    def __init__(self, config):
+        super(TronConfig, self).__init__(config)
+
+    def get_cluster_name(self):
+        """:returns The name of the Tron cluster"""
+        try:
+            return self['cluster_name']
+        except KeyError:
+            raise TronNotConfigured('Could not find name of Tron cluster in system Tron config')
+
+    def get_default_paasta_cluster(self):
+        """:returns The PaaSTA cluster to run actions on by default"""
+        try:
+            return self['default_paasta_cluster']
+        except KeyError:
+            raise TronNotConfigured('Could not find default PaaSTA cluster in system Tron config')
+
+    def get_url(self):
+        """:returns The URL for the Tron master's API"""
+        try:
+            return self['url']
+        except KeyError:
+            raise TronNotConfigured('Could not find URL of Tron master in system Tron config')
+
+
+def load_tron_config():
+    return TronConfig(load_system_paasta_config().get_tron_config())
+
+
+class TronClient:
+
+    def __init__(self, url):
+        self.master_url = url
+
+    def _request(self, method, url, data):
+        headers = {'User-Agent': get_user_agent()}
+        kwargs = {
+            'url': urljoin(self.master_url, url),
+            'headers': headers,
+        }
+        if method == 'GET':
+            kwargs['params'] = data
+            response = requests.get(**kwargs)
+        elif method == 'POST':
+            kwargs['data'] = data
+            response = requests.post(**kwargs)
+        else:
+            raise ValueError('Unrecognized method: {}'.format(method))
+
+        # Raise an exception if unsuccessful.
+        response.raise_for_status()
+        return response.json()
+
+    def _get(self, url, data=None):
+        return self._request('GET', url, data)
+
+    def _post(self, url, data=None):
+        return self._request('POST', url, data)
+
+    def update_namespace(self, namespace, new_config):
+        current_config = self._get('/api/config', {'name': namespace})
```

Comment:
seems like this validation should be moved up a layer (what if I want to update it in spite of it being the same?).